File: frontend/src/types/realms.ts
import type { Actor } from "./actor";
import type { Aggregate } from "./aggregate";
import type { Change } from "./change";
import type { CustomAttribute } from "./custom";
import type { PasswordSettings, UniqueNameSettings } from "./settings";
import type { SearchPayload, SortOption } from "./search";

export type CreateOrReplaceRealmPayload = {
  uniqueSlug: string;
  displayName?: string;
  description?: string;
  url?: string;
  uniqueNameSettings: UniqueNameSettings;
  passwordSettings: PasswordSettings;
  requireUniqueEmail: boolean;
  requireConfirmedAccount: boolean;
  customAttributes: CustomAttribute[];
};

export type Realm = Aggregate & {
  uniqueSlug: string;
  displayName?: string | null;
  description?: string | null;
  secretChangedBy: Actor;
  secretChangedOn: string;
  url?: string | null;
  uniqueNameSettings: UniqueNameSettings;
  passwordSettings: PasswordSettings;
  requireUniqueEmail: boolean;
  requireConfirmedAccount: boolean;
  customAttributes: CustomAttribute[];
};

export type RealmSort = "CreatedOn" | "DisplayName" | "SecretChangedOn" | "UniqueSlug" | "UpdatedOn";

export type RealmSortOption = SortOption & {
  field: RealmSort;
};

export type SearchRealmsPayload = SearchPayload & {
  sort: RealmSortOption[];
};

export type UpdateRealmPayload = {
  uniqueSlug?: string;
  displayName?: Change<string>;
  description?: Change<string>;
  secret?: Change<string>;
  url?: Change<string>;
  uniqueNameSettings?: UniqueNameSettings;
  passwordSettings?: PasswordSettings;
  requireUniqueEmail?: boolean;
  requireConfirmedAccount?: boolean;
  customAttributes: CustomAttribute[];
};

File: backend/src/Krakenar.EntityFrameworkCore.Relational/KrakenarDb/Realms.cs
using Krakenar.EntityFrameworkCore.Relational.Entities;
using Logitar.Data;

namespace Krakenar.EntityFrameworkCore.Relational.KrakenarDb;

public static class Realms
{
  public static readonly TableId Table = new(Schemas.Krakenar, nameof(KrakenarContext.Realms), alias: null);

  public static readonly ColumnId CreatedBy = new(nameof(Realm.CreatedBy), Table);
  public static readonly ColumnId CreatedOn = new(nameof(Realm.CreatedOn), Table);
  public static readonly ColumnId StreamId = new(nameof(Realm.StreamId), Table);
  public static readonly ColumnId UpdatedBy = new(nameof(Realm.UpdatedBy), Table);
  public static readonly ColumnId UpdatedOn = new(nameof(Realm.UpdatedOn), Table);
  public static readonly ColumnId Version = new(nameof(Realm.Version), Table);

  public static readonly ColumnId AllowedUniqueNameCharacters = new(nameof(Realm.AllowedUniqueNameCharacters), Table);
  public static readonly ColumnId CustomAttributes = new(nameof(Realm.CustomAttributes), Table);
  public static readonly ColumnId Description = new(nameof(Realm.Description), Table);
  public static readonly ColumnId DisplayName = new(nameof(Realm.DisplayName), Table);
  public static readonly ColumnId Id = new(nameof(Realm.Id), Table);
  public static readonly ColumnId PasswordHashingStrategy = new(nameof(Realm.PasswordHashingStrategy), Table);
  public static readonly ColumnId PasswordsRequireDigit = new(nameof(Realm.PasswordsRequireDigit), Table);
  public static readonly ColumnId PasswordsRequireLowercase = new(nameof(Realm.PasswordsRequireLowercase), Table);
  public static readonly ColumnId PasswordsRequireNonAlphanumeric = new(nameof(Realm.PasswordsRequireNonAlphanumeric), Table);
  public static readonly ColumnId PasswordsRequireUppercase = new(nameof(Realm.PasswordsRequireUppercase), Table);
  public static readonly ColumnId RealmId = new(nameof(Realm.RealmId), Table);
  public static readonly ColumnId RequireConfirmedAccount = new(nameof(Realm.RequireConfirmedAccount), Table);
  public static readonly ColumnId RequiredPasswordLength = new(nameof(Realm.RequiredPasswordLength), Table);
  public static readonly ColumnId RequiredPasswordUniqueChars = new(nameof(Realm.RequiredPasswordUniqueChars), Table);
  public static readonly ColumnId RequireUniqueEmail = new(nameof(Realm.RequireUniqueEmail), Table);
  public static readonly ColumnId Secret = new(nameof(Realm.Secret), Table);
  public static readonly ColumnId SecretChangedBy = new(nameof(Realm.SecretChangedBy), Table);
  public static readonly ColumnId SecretChangedOn = new(nameof(Realm.SecretChangedOn), Table);
  public static readonly ColumnId UniqueSlug = new(nameof(Realm.UniqueSlug), Table);
  public static readonly ColumnId UniqueSlugNormalized = new(nameof(Realm.UniqueSlugNormalized), Table);
  public static readonly ColumnId Url = new(nameof(Realm.Url), Table);
}
